assets/js/components/product-control/index.js

```diff
@@ -1,54 +1,237 @@
 /**
  * External dependencies
  */
-import { __ } from '@wordpress/i18n';
+import { __, _n, sprintf } from '@wordpress/i18n';
 import { Component, Fragment } from '@wordpress/element';
-import { debounce, find } from 'lodash';
+import { addQueryArgs } from '@wordpress/url';
+import apiFetch from '@wordpress/api-fetch';
+import { debounce, find, escapeRegExp, isEmpty } from 'lodash';
 import PropTypes from 'prop-types';
-import { SearchListControl } from '@woocommerce/components';
+import {
+	SearchListControl,
+	SearchListItem,
+} from '@woocommerce/components';
+import { Spinner, MenuItem } from '@wordpress/components';
+import classnames from 'classnames';
 
 /**
  * Internal dependencies
  */
 import { isLargeCatalog, getProducts } from '../utils';
+import {
+	IconRadioSelected,
+	IconRadioUnselected,
+} from '../icons';
+import './style.scss';
+
+function getHighlightedName( name, search ) {
+	if ( ! search ) {
+		return name;
+	}
+	const re = new RegExp( escapeRegExp( search ), 'ig' );
+	return name.replace( re, '<strong>$&</strong>' );
+}
+
+const getInteractionIcon = ( isSelected = false ) => {
+	return isSelected ? <IconRadioSelected /> : <IconRadioUnselected />;
+};
 
 class ProductControl extends Component {
 	constructor() {
 		super( ...arguments );
 		this.state = {
-			list: [],
+			products: [],
+			product: 0,
+			variationsList: {},
+			variationsLoading: false,
 			loading: true,
 		};
 
 		this.debouncedOnSearch = debounce( this.onSearch.bind( this ), 400 );
+		this.debouncedGetVariations = debounce( this.getVariations.bind( this ), 200 );
+		this.renderItem = this.renderItem.bind( this );
+		this.onProductSelect = this.onProductSelect.bind( this );
 	}
 
 	componentDidMount() {
 		const { selected } = this.props;
 
 		getProducts( { selected } )
-			.then( ( list ) => {
-				this.setState( { list, loading: false } );
+			.then( ( products ) => {
+				products = products.map( ( product ) => {
+					const count = product.variations ? product.variations.length : 0;
+					return {
+						...product,
+						parent: 0,
+						count: count,
+					};
+				} );
+				this.setState( { products, loading: false } );
 			} )
 			.catch( () => {
-				this.setState( { list: [], loading: false } );
+				this.setState( { products: [], loading: false } );
+			} );
+	}
+
+	componentDidUpdate( prevProps, prevState ) {
+		if ( prevState.product !== this.state.product ) {
+			this.debouncedGetVariations();
+		}
+	}
+
+	getVariations() {
+		const { product, variationsList } = this.state;
+		if ( ! product ) {
+			this.setState( {
+				variationsList: {},
+				variationsLoading: false,
+			} );
+			return;
+		}
+		if ( ! variationsList[ product ] ) {
```

Comment:
I noticed we are trying to load variations for all products, even the ones that don't have any. What about adding here something like this check to prevent making API calls to the variations endpoint for products we know have no variations?
```JS

		const productDetails = this.state.products.find( ( prod ) => prod.id === product );
		if ( ! productDetails.variations || productDetails.variations.length === 0 ) {
			return;
		}
```